File: AEDs/AEDs_II/TrabalhosPraticos/TP01/TP01Q12-PalindromoRec/palindromo.c
#include <biblioteca_c.h>

// clear && gcc palindromo.c -o palindromo && ./palindromo < pub.in > out.txt

bool CaractereEspecial(char c) {
	return !(isalpha(c) || c == ' ' || c == '-' || c >= '0' && c <= '9');
}

bool isPalindromo(char* string, int comeco, int fim) {
	if (comeco >= fim) {
		return true;
	}

	if (CaractereEspecial(string[comeco])) {
		if (string[comeco] != string[fim - 1]) {
			return false;
		} else if (string[comeco + 1] != string[fim]) {
			return false;
		} else {
			return isPalindromo(string, comeco + 2, fim - 2);
		}
	} else {
		if (string[comeco] != string[fim]) {
			return false;
		}
	}

	return isPalindromo(string, ++comeco, --fim);
}

int main() {

	char* string = getstr(0);

	while (strcmp(string, "FIM")) {
		if (isPalindromo(string, 0, strlen(string) - 1)) {
			printf("SIM\n");
		} else {
			printf("NAO\n");
		}
		strcopy(&string, getstr(0));
	}

	return 0;
}

File: AEDs/AEDs_II_Deprecated/TrabalhosPraticos/TP01/TP01Q11-PalindromoRec/palindromo.java
import java.util.Scanner;

// clear && javac palindromo.java && java palindromo < pub.in > out.txt

public class palindromo {
	public static boolean isPalindromo(String string, int comeco, int fim) {
		if (comeco >= fim) {
			return true;
		}

		if (string.charAt(comeco) != string.charAt(fim)) {
			return false;
		}

		return isPalindromo(string, ++comeco, --fim);
	}

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);

		String string = scanner.nextLine();

		while (!string.equals("FIM")) {
			if (isPalindromo(string, 0, string.length() - 1)) {
				System.out.println("SIM");
			} else {
				System.out.println("NAO");
			}

			string = scanner.nextLine();
		}

		scanner.close();
	}
}